src/platform/plugins/shared/chart_expressions/expression_partition_vis/public/utils/layers/get_color.ts

```diff
@@ -11,60 +11,42 @@ import { ArrayNode } from '@elastic/charts';
 import { isEqual } from 'lodash';
 import type { PaletteRegistry, SeriesLayer, PaletteOutput, PaletteDefinition } from '@kbn/coloring';
 import type { FieldFormatsStart } from '@kbn/field-formats-plugin/public';
-import type { FieldFormat } from '@kbn/field-formats-plugin/common';
 import { lightenColor } from '@kbn/charts-plugin/public';
-import type { Datatable } from '@kbn/expressions-plugin/public';
 import { BucketColumns, ChartTypes, PartitionVisParams } from '../../../common/types';
 import { DistinctSeries } from '../get_distinct_series';
-import { getNodeLabel } from './get_node_labels';
 
 const isTreemapOrMosaicChart = (shape: ChartTypes) =>
   [ChartTypes.MOSAIC, ChartTypes.TREEMAP].includes(shape);
 
 export const byDataColorPaletteMap = (
-  rows: Datatable['rows'],
-  column: Partial<BucketColumns>,
   paletteDefinition: PaletteDefinition,
   { params }: PaletteOutput,
-  formatters: Record<string, FieldFormat | undefined>,
-  formatter: FieldFormatsStart
+  colorIndexMap: Map<string, number>
 ) => {
-  const colorMap = new Map<string, string | undefined>(
-    rows.map((item) => {
-      const formattedName = getNodeLabel(
-        item[column.id ?? ''],
-        column,
-        formatters,
-        formatter.deserialize
-      );
-      return [formattedName, undefined];
-    })
-  );
-  let rankAtDepth = 0;
+  const colorCache = new Map<string, string | undefined>();
 
   return {
-    getColor: (item: unknown) => {
+    getColor: (item: string) => {
       const key = String(item);
-      if (!colorMap.has(key)) return;
+      let color = colorCache.get(key);
 
-      let color = colorMap.get(key);
-      if (color) {
-        return color;
-      }
+      if (color) return color;
+
+      const colorIndex = colorIndexMap.get(key) ?? -1;
       color =
         paletteDefinition.getCategoricalColor(
           [
             {
               name: key,
-              totalSeriesAtDepth: colorMap.size,
-              rankAtDepth: rankAtDepth++,
+              totalSeriesAtDepth: colorIndexMap.size,
+              rankAtDepth: colorIndex,
```

Comment:
The only thing I've noticed is that there is a difference in the color depth used in treemap when using the legacy vs the new color mapping. In this case, the legacy color is not providing darker colors for lower depth level, where instead the color mapping is doing it. In pie chart the behaviour looks fine, so it looks only an issue with treemap.
See screenshots:

<img width="1477" alt="Screenshot 2025-02-19 at 10 32 05" src="https://github.com/user-attachments/assets/e4be6c79-d234-4fe9-be26-9180b0eb2336" />

<img width="1529" alt="Screenshot 2025-02-19 at 10 32 01" src="https://github.com/user-attachments/assets/6d4b0fe0-4a03-40cf-8cf5-6fafc14c603e" />

Reply:
Fixed in af447da06a020111185b2d359d9b10fd7d188cb0. Treemap with 2+ layers now match between legacy and color mapping logic.

![Zight Recording 2025-03-03 at 08 22 24 PM](https://github.com/user-attachments/assets/cd125a0d-7bd2-4e3d-a35c-232c7b30e0ec)